mgr: override OSD hostname with CRUSH physical host in all metadata paths

senolcolak opened the pull request:
## Problem

In containerized Ceph deployments (e.g. Rook/OCS), OSD daemons report the
pod or container hostname. The CRUSH map records the physical host. This
mismatch causes `ceph device ls` to show pod names instead of physical
hosts, breaking host-based device correlation.

Fixes: https://tracker.ceph.com/issues/73080

## What was wrong with the previous version

The original patch inserted `_override_osd_crush_hostname()` inside
`if (m->service_daemon)` in `handle_open`. OSDs connect with
`service_daemon = false`, so that branch is never reached for OSD daemons —
the override was dead code.

In addition, the `handle_update` insertion called `with_osdmap` (which
acquires `Objecter::rwlock`) while holding `state->lock`, introducing a
new lock ordering `state->lock → Objecter::rwlock` not present elsewhere.
And setting `state->hostname` directly before `update_metadata` would
corrupt the `by_server` index, because `_erase` reads `dm->hostname` at
call time to locate the entry to remove.

## Solution

Move the override to the three paths where OSD metadata actually flows:

1. **`MetadataUpdate::finish`** — async mon metadata fetch (primary path,
   runs in Ceph finisher thread). CRUSH resolution happens before taking
   `state->lock` to preserve lock ordering. The CRUSH hostname is injected
   via `m["hostname"]` so `DaemonStateIndex::update_metadata` / `set_metadata`
   picks it up correctly, keeping the `by_server` index consistent.

2. **`Mgr::load_all_metadata`** — OSD bulk-load at startup/failover.

3. All five `MetadataUpdate` construction sites are updated to pass
   `cluster_state` to the new 3-argument constructor.

A static free function `crush_hostname_for_osd(ClusterState&, int)` in
`Mgr.cc` calls `cluster_state.with_osdmap` and returns the CRUSH `"host"`
bucket name for the OSD, or an empty string when absent. The override is a
no-op when the OSD is absent from CRUSH or the OSD map is not yet loaded.

## Testing

Verify in a Rook/containerized environment:
1. `ceph device ls` shows physical hostnames, not pod names
2. `ceph osd metadata <id>` shows correct `hostname` field
3. No regression on bare-metal (CRUSH host = reported hostname → no-op)
4. mgr failover correctly restores hostnames via `load_all_metadata`

ceph-jenkins commented:
Can one of the admins verify this patch?

ceph-jenkins commented:
Can one of the admins verify this patch?

ceph-jenkins commented:
Can one of the admins verify this patch?

rzarzynski reviewed (approved):
Overall LGTM with a nit. Please see comments below.

senolcolak commented on src/mgr/DaemonServer.cc:
```diff
@@ -649,7 +665,23 @@ bool DaemonServer::handle_update(const ref_t<MMgrUpdate>& m)
+		}
+	      }
+	    });
+	  } catch (const std::exception&) {
```
```suggestion
      } catch (const std::exception& e) {
        dout(20) << "DaemonServer: cannot derive CRUSH hostname for "
                 << key.type << "." << key.name
                 << " from OSDMap: " << e.what() << dendl;
      }
```

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok to test" to allow the Jenkins jobs to run.

SrinivasaBharath commented:
@senolcolak  -If this PR requires re-testing, please add the needs-qa label accordingly.

senolcolak commented:
Hi @rzarzynski 
all checks are green now. Is this ready to merge, or is anything else needed from my side?

senolcolak committed:
mgr: override OSD hostname with CRUSH physical host in all metadata paths

In containerized deployments (e.g. Rook), OSD daemons report the
pod/container hostname. The CRUSH map records the physical host.
This mismatch causes 'ceph device ls' to show pod names instead of
physical hosts, breaking host-based device correlation.

Fix by deriving the CRUSH host bucket name for each OSD and substituting
it as the canonical hostname in all three metadata paths:
- MetadataUpdate::finish (async mon metadata fetch, primary path)
- Mgr::load_all_metadata OSD bulk-load (startup/failover)
- DaemonServer::fetch_missing_metadata (via MetadataUpdate)

The override is a no-op when the OSD is absent from CRUSH or when the
OSD map is not yet available.

Fixes: https://tracker.ceph.com/issues/73080
Signed-off-by: senol.colak <senol.colak@sap.com>

senolcolak commented:
this patch needs to run tests against  gardener shoots.

senolcolak commented on src/mgr/DaemonServer.cc:
```diff
@@ -565,7 +565,23 @@ bool DaemonServer::handle_open(const ref_t<MMgrOpen>& m)
+	      }
+	    }
+	  });
+	} catch (const std::exception&) {
```
Fixed in the updated patch. The rework moves the override out of `DaemonServer` entirely — `handle_open` and `handle_update` are reverted to their originals. The CRUSH hostname resolution now lives in `MetadataUpdate::finish` (in `Mgr.cc`) where both catch blocks use `dout(5)` to log failures.

zoorpha reviewed (commented):
**Overall: LGTM** — the v2 rework is a significant improvement over v1. The dead-code issue is fixed, lock ordering is preserved, and the `by_server` index consistency concern from the previous iteration is properly addressed. The approach is sound and the comments explaining the locking rationale are excellent.

I have a few minor suggestions below. Nothing blocking, but worth addressing before merge.

---

### 1. Redundant `state->hostname` assignment in the new-daemon path

In `MetadataUpdate::finish()`, the `!daemon_state.exists(key)` branch sets `state->hostname` three times:

```cpp
state->hostname = daemon_meta.at("hostname").get_str();  // reported hostname
// ...
state->hostname = crush_host;                            // CRUSH override
// ...
state->set_metadata(m);                                  // sets it again from m
```

Because `DaemonState::set_metadata()` updates `hostname` from the map when `"hostname"` is present, the middle assignment is redundant in the OSD-with-CRUSH-host case. Consider simplifying to:

```cpp
// For non-OSDs or OSDs without a CRUSH host, preserve the reported hostname
if (key.type != "osd" || crush_host.empty()) {
    m["hostname"] = daemon_meta.at("hostname").get_str();
} else {
    m["hostname"] = crush_host;
}
state->set_metadata(m);
```

This removes the direct `state->hostname` assignments entirely and lets `set_metadata` own the field.

---

### 2. Add automated test coverage

The PR description lists manual Rook verification steps, which is great for the final check, but core mgr metadata indexing changes should have some automated coverage. Two suggestions:

**A. Unit test for `crush_hostname_for_osd` and `by_server` consistency**

The `crush_hostname_for_osd` helper is currently `static` in `Mgr.cc`, so it's invisible to `src/test/mgr/`. Could you either:
- Move it to `ClusterState` as a public method (e.g., `ClusterState::get_crush_hostname_for_osd(int)`), or
- Declare it in `Mgr.h` and make it non-static?

Then we can add a `unittest_mgr_clusterstate` test that builds a `CrushWrapper`, injects it into an `OSDMap`, and asserts the helper returns the correct `"host"` bucket. We should also test that `DaemonStateIndex::update_metadata` correctly migrates the `by_server` index when the hostname changes.

**B. Integration test without containers**

You don't need a real Rook cluster to verify this. A simple vstart or teuthology test can manufacture the mismatch by moving an OSD to a synthetic CRUSH host:

```bash
ceph osd crush move osd.0 host=fake-container-name
# ... verify ceph osd metadata 0 shows hostname=fake-container-name
```

This exercises the exact code path (CRUSH host ≠ reported hostname) without any container infrastructure. I'd suggest adding either:
- A shell workunit under `qa/workunits/mgr/`, or
- A Python teuthology task under `qa/tasks/mgr/` that performs the CRUSH move, fails the mgr, and asserts `ceph osd metadata 0 | jq -r '.hostname'`.

---

### 3. Nit: `load_all_metadata` `dout(4)` behavior change

Not a bug, just an FYI: in `load_all_metadata`, the log line now prints the post-override hostname rather than the raw reported hostname. This is the correct behavior, but if anyone is debugging metadata issues they may notice the difference after this patch.

---

**Thanks for the careful rework — the locking comments in `MetadataUpdate::finish` are particularly good and will save future maintainers a lot of head-scratching.**

senolcolak committed:
mgr: preserve reported hostname when CRUSH host is unavailable

The previous patch only injected hostname into the metadata map for OSDs
that were placed under a CRUSH host bucket. For mds/mgr/mon daemons and
for OSDs absent from CRUSH, the original 'state->hostname = ...' line
was removed without a replacement, so update_metadata received a map
with no 'hostname' key. DaemonState::set_metadata only updates
state->hostname when the map carries it, so a daemon restarting on a
different host would be re-inserted in by_server under its stale
hostname.

Always seed m["hostname"] from the metadata-reported hostname, then
prefer the CRUSH host when available. Apply the same shape to the
new-daemon branch so set_metadata uniformly drives state->hostname,
removing the redundant direct assignment.

Reported-by: code review on PR #66757
Signed-off-by: senol.colak <senol.colak@sap.com>

senolcolak commented:
Hi @rzarzynski, @adk3798, @SrinivasaBharath — small follow-up commit on top of the previously approved `dcd9c5b`:

- `ec59888` — `src/mgr/Mgr.cc` only, +11/-10

It closes a `by_server` index gap in the v2 patch: `daemon_meta.erase("hostname")` was unconditional, but the re-insertion only happened when a CRUSH `host` bucket was found. For mds/mgr/mon daemons and for OSDs not placed under a `host` bucket, `m` ended up without a `hostname` key, so `DaemonState::set_metadata()` left `state->hostname` untouched and `_insert` re-keyed `by_server` under a stale value. Fix: `m["hostname"] = !crush_host.empty() ? crush_host : reported_hostname`. Lock ordering and OSD-with-CRUSH-host behavior are unchanged.

Since QA last ran against `dcd9c5b`, could a maintainer add the `needs-qa` label so the metadata path gets re-validated against `ec59888`?

Thanks for the patience on this one.

rzarzynski commented:
Let's retest it!

shraddhaag commented:
RADOS Approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues77627

SrinivasaBharath commented:
jenkins test make check arm64
